File: BOJ/1759_암호만들기.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#define endl '\n'
using namespace std;

vector<char> alp;
vector<bool> visited;
vector<char> res;

char vowels[5] = { 'a', 'e', 'i', 'o','u' };

bool promise() {
	int curVowels = 0, curConsonants = 0;
	for (int i = 0; i < res.size(); i++) {
		bool v = false;
		for (int j = 0; j < 5; j++) {
			if (res[i] == vowels[j]) v = true;
		}
		v ? curVowels++ : curConsonants++;
	}

	return ( curVowels >= 1 && curConsonants >= 2 );
}
void dfs(int length, int l, int c) {
	if (length == l) {
		if (promise()) {
			for (int i = 0; i < l; i++) {
				cout << res[i];
			}cout << endl;
		}
		return;
	}

	for (int i = 0; i < c; i++) {
		if (!visited[i]) {
			if (length == 0 || (length != 0 && res[length - 1] < alp[i])) {
				visited[i] = true;
				res[length] = alp[i];
				dfs(length + 1, l, c);
				visited[i] = false;
			}
		}
	}

}

int main() {
	int l, c;
	cin >> l >> c;
	alp = vector<char>(c);
	res = vector<char>(l);
	visited = vector<bool>(c, false);


	for (int i = 0; i < c; i++) cin >> alp[i];
	sort(alp.begin(), alp.end());
	dfs(0, l, c);

	return 0;
}

File: BOJ/1759_암호만들기.java
import java.io.*;
import java.util.*;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer tk;
	static StringBuilder sb = new StringBuilder();
	static int l, c;
	static char[] ch;
	static char[] arr;
	static boolean[] visited;
	
	static void dfs(int len, int vowels, int consonants) {
		if(len == l) {
			if(vowels >= 1 && consonants >= 2) {
				for(int i=0; i<len; i++) sb.append(arr[i]);
				sb.append("\n");
			}
			return;
		}
		
		for(int i=0; i<c; i++) {
			if(!visited[i] && (len == 0 || arr[len-1] < ch[i])) {
				visited[i] = true;
				arr[len] = ch[i];
				if(ch[i] == 'a' || ch[i] == 'e' || ch[i] == 'i' || ch[i] == 'o' || ch[i] == 'u') 
					dfs(len+1, vowels+1, consonants);
				else dfs(len+1, vowels, consonants+1);
				visited[i] = false;
			}
		}
	}
	
	public static void main(String[] args) throws IOException{
		tk = new StringTokenizer(br.readLine());
		l = Integer.parseInt(tk.nextToken());
		c = Integer.parseInt(tk.nextToken());

		ch = new char[c];
		arr = new char[l];
		visited = new boolean[c];
		
		tk = new StringTokenizer(br.readLine());
		for(int i=0; i<c; i++) ch[i] = tk.nextToken().charAt(0);
		Arrays.sort(ch);
		
		dfs(0, 0, 0);
		System.out.println(sb.toString());
	}

}

